### ISSUE TITLE
is there a plan to support 3d view 

### ISSUE THREAD

**Issue by giswrok** [openlayers/openlayers#14083]
Is there a plan to support 3D views similar to mapbox GL? I really don't want to use another API to do the existing functions again;

**Comment by Rskut** [comment]
https://openlayers.org/ol-cesium/

**Comment by carlosjpr-collab** [comment]
@Rskut  thank, but we need a real 3d support not as plugin.

**Comment by gtnbssn** [comment]
Curious about this as well. The ol-cesium examples and documentation are giving me the impression to be outdated.

My impression is that what people want isn't a globe (which is what cesium does I think?) but webmercator on a plane in a WebGL canvas with controls to move a camera around, moving the plane's vertices up and down according to dem data to show relief, and possibly other bells and whistles like extruded buildings, etc.

With WebGPU now being around the corner, maybe it would be possible to skip WebGL implementation and go straight to the newer API.

Any that's all wishful thinking and I don't understand all the implications, but it sure would be nice to have™.

**Comment by roshomon** [comment]
3d view or globe view support in latest version?

**Comment by ahocevar** [comment]
For the time being, there is no plan to support a 3d globe view. 2.5d to show e.g. extruded buildings is something we might be able to support in the future.

**Comment by giswrok** [comment]
If OL can support 2.5d, it is already sufficient in many scenarios. Looking forward to it very much, thank you for your efforts







--

浙江臻善科技股份有限公司——薛晓龙




At 2024-01-24 17:33:54, "Andreas Hocevar" ***@***.***> wrote:

For the time being, there is no plan to support a 3d globe view. 2.5d to show e.g. extruded buildings is something we might be able to support in the future.

**Comment by silvester-pari** [comment]
+1 on this topic in general, and contrary to gtnbssn's impression that a 2/2.5 plane map is what the majority of users want in this regard, I rather have the impression what's wanted is a globe to show large-scale/global data (or data at the poles), as is common nowadays in map interfaces like MapLibre (since this year), Mapbox, Google maps, Apple maps etc.

It was interesting to read the discussion in the MapLibre world in https://github.com/maplibre/maplibre/discussions/161, and also very interesting to see the first experiments here with OpenLayers in https://github.com/openlayers/openlayers/discussions/16109.

We are currently struggling to find a good approach for "3d" globes/maps compatible with OpenLayers - ol-cesium seems to be currently very limited/not very performant, and building our tech stack from scratch with e.g. vanilla Cesium, OpenGlobus or MapLibre is not something we are going to do anytime soon.

So... any momentum in this direction would be very nice to see! 👍

**Comment by ahocevar** [comment]
@silvester-pari It is possible to create OffscreenCanvas tiles from whatever you currently have configured for an OpenLayers map, and feed these tiles into OpenGlobus. This way, you only configure the camera in OpenGlobus, and get the same map as 3D globe that you have configured for the 2D map.

If you're interested in that, talk to @RobertOrthofer - he's currently working on this in an application. My ultimate goal is to make such an OffscreenCanvas tiles engine a first-class citizen in OpenLayers, because it would also allow for performance optimized vector tiles created from ol-mapbox-style.

**Comment by jahow** [comment]
@ahocevar although I've always been in favor of a "texture provider" approach where OpenLayers does not do 3D but offers the right APIs to load data in a 3d globe, we also have to recognize that this is subject to limitations:

1. If OpenLayers only provides images, either tiles or full coverages (e.g. WMS), this means no floating labels on top of the terrain and no actual vector rendering of features (which leads to stretching/visibility issues depending on the terrain and view angle)
2. Style properties will not work as good as in 2D: a width pixel value wouldn't mean much once projected on a 3D surface
3. Loading tiles of various scales to account for tilted views would probably be awkward with the current API
4. Having two libraries essentially run in parallel (OL and the globe one) will most likely leave us with a glass ceiling when it comes to performance

Maybe we could come up with a future-proof plan to properly handle 3D, now that essentially _every other library_ does it? I don't have any strong opinion on this, just thinking out loud.

**Comment by gberaudo** [comment]
Hi @silvester-pari , I am the maintainer of OL-Cesium.

Together with others from the OpenLayers community, notably @ahocevar , we created OL-Cesium to bring 3D to OpenLayers, it was 10 years ago.
OL-Cesium introspects your OL layers and transform them to Cesium equivalents.
OL-Cesium tracks view/camera changes and update the value in the other library.
It inherits limitations from both OpenLayers and CesiumJS, notably regarding internal states, or limited tilegrid and projections support. However I think it has a great architecture, is valuable, and is open to contributions. It is a great asset of the OpenLayers community.

Lately 3D technologies and visualization has gained a lot of traction, I think it is time for a discussion about what works, what does not work and where we are heading. What do you think @ahocevar ?

**Comment by silvester-pari** [comment]
Hi @gberaudo and thank you for your work on ol-cesium! For basic support (xyz layers etc.) it seems to work great, however we have lots of advanced usage of OL (GeoTIFF sources, WMTS layer from capabilities, STAC layer groups based on ol-stac, flatgeobufs, Mapbox style layers etc.) and many of these things would need to be implemented additionally in order to be supported on a Cesium globe. In an ideal world, all the work we put into displaying thins on the 2D OL map would easily be available in 3D without too much extra effort.
The solution mentioned by @ahocevar (using OL as a renderer and mapping it on the globe via canvas tiles) seems like a good (temporary) compromise, as it would basically allow to display anything rendered by OL on a globe, of course with the limitations mentioned by @jahow, plus no interactions, plus no support for WebGL stuff I guess. Ultimately, it would be awesome if OL would natively support a 3D view, so no second library is required and everything that can be rendered in 2D can also be rendered in 3D, with ideally some basic interaction support (hover, click, maybe even basic drawing support although that's hardly usable in most 3D globe libraries).

**Comment by andreaordonselli** [comment]
The boys from Oslandia created Giro3D https://gitlab.com/giro3d/giro3d
I don't think it's easy to bring an entire WebGL 3D Globe ecosystem integrated into OpenLayers, so "Giro3D" (with three.js) seems like a great opportunity, which does not rewrite OL styles and uses a drape applied to the ground (the same thing suggested above).
As far as I'm concerned, it's better than integration with Cesium, which when it has to become serious, with a tool like 3Dmeasure or other, becomes paid.
I am already collaborating with Oslandia to integrate Giro3D in the OpenLayers export of qgis2web (I am the maintenance technician).
Now I'm looking for:
- 3d globe
- 3d terrain open api (https://terrain.reearth.land/ ?)
- 3d buildings open api (openfreemap?)
- sun and shadows
- 3d measure
- import 3d tiles from library
